[データサイエンスのためのPython入門13〜DataFrameのフィルタ操作の基本\(超重要\)〜](https://datawokagaku.com/dataframe_filter/)

こんにちは！勉強してますか？

さて、かめ@米国データサイエンティストさんご提供の「データサイエンスのためのPython入門」ですが、今日はですね、第13回、題して「DataFrameのフィルタ操作の基本(超重要)」編をやっていきたいと思います！

今日もDataFrameに対するフィルタ操作をみていきましょう〜〜〜。

さて、いつものやつ〜〜〜。

In [19]:
import pandas as pd

今回も、前回同様タイタニックのデータを使いま〜す。

In [20]:
df = pd.read_csv('data/titanic/train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

さて、この891レコードから、生存者のレコードのみを抽出してみましょう。

まずは、カラムが`Survived`なSeriesを取り出してみましょう。

In [27]:
df['Survived']

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

死亡は`0`、生存は`1`で記録されているようですね。

今度は次のようにしてみます。

In [29]:
df['Survived'] == 1

0      False
1       True
2       True
3       True
4      False
       ...  
886    False
887     True
888    False
889     True
890    False
Name: Survived, Length: 891, dtype: bool

カラムの値が、`0`/`1`から`True`/`False`に変わりましたね。どういうことなんでしょうか？

In [36]:
print(type(df['Survived']))
print(type(df['Survived'] == 1))

<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>


同じ型が返ってきますね。きっと、`==`演算子をオーバーロードして、レシーバーのSeriesの指定されたカラムを一個ずつ検証していって`True`/`False`のSeriesとして返すみたいな感じなんでしょうね。

さて、その`df['Survived'] == 1`を`df[]`に渡してみると……

In [37]:
df[df['Survived'] == 1]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[342 rows x 12 columns]

`Survived`が`1`のデータだけが返ってきた〜〜〜！不思議な挙動だなあ。きっと、`df['Survived'] == 1`には、`Survived`カラムをフィルタしてね、っていう情報も入ってるんでしょうね〜〜〜。

ここはやっぱり、生存者のみを抽出したこのDataFrameの統計量を見てみくなりますよね！そんなときは……

In [40]:
df[df['Survived'] == 1].describe()

PassengerId  Survived      Pclass         Age       SibSp       Parch  \
count   342.000000     342.0  342.000000  290.000000  342.000000  342.000000   
mean    444.368421       1.0    1.950292   28.343690    0.473684    0.464912   
std     252.358840       0.0    0.863321   14.950952    0.708688    0.771712   
min       2.000000       1.0    1.000000    0.420000    0.000000    0.000000   
25%     250.750000       1.0    1.000000   19.000000    0.000000    0.000000   
50%     439.500000       1.0    2.000000   28.000000    0.000000    0.000000   
75%     651.500000       1.0    3.000000   36.000000    1.000000    1.000000   
max     890.000000       1.0    3.000000   80.000000    4.000000    5.000000   

             Fare  
count  342.000000  
mean    48.395408  
std     66.596998  
min      0.000000  
25%     12.475000  
50%     26.000000  
75%     57.000000  
max    512.329200

ジャジャーン！元のDataFrameと比較してみましょう。

In [41]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

`Fare`（料金）の平均値（`mean`）を見ると、生存者のほうがちょっと高いですね……。これが意味するところは……と興味がつきないところですが、この差異って本当に差があるといえるの？ってことについては、かめさんが統計学入門講座で解き明かしてくれるそうです！楽しみですね！

さて、他にもいろんな条件でフィルタしてみますか〜〜〜。

In [ ]:
df[df['Age'] >= 60]
df[df['Pclass'] == 1]
df[df['Sex'] == 'female']

こんな感じでいろいろ抽出できちゃいます！

さて、こうなってくると今度は、抽出条件をあれこれ組み合わせたくなりますね！

Python（やSQLとか）だと`and`や`or`を使うところですが、DataFrameのフィルタ操作は`&`と`|`を使うんですね〜。注意です！

60歳以上の女性の場合はこんな感じ。

In [57]:
df[(df['Age'] >= 60) & (df['Sex'] == 'female')]

PassengerId  Survived  Pclass  \
275          276         1       1   
366          367         1       1   
483          484         1       3   
829          830         1       1   

                                                 Name     Sex   Age  SibSp  \
275                 Andrews, Miss. Kornelia Theodosia  female  63.0      1   
366  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0      1   
483                            Turkula, Mrs. (Hedwig)  female  63.0      0   
829         Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   

     Parch  Ticket     Fare Cabin Embarked  
275      0   13502  77.9583    D7        S  
366      0  110813  75.2500   D37        C  
483      0    4134   9.5875   NaN        S  
829      0  113572  80.0000   B28      NaN

1stクラスの乗客、もしくは、10歳未満の場合！

In [58]:
df[(df['Pclass'] == 1) | (df['Age'] < 10)]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
7              8         0       3   
10            11         1       3   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
7        1    349909  21.0750          NaN        S  
10       1   PP 9549  16.7000           G6        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[275 rows x 12 columns]

いっぱいでてきたな……。

さて、いままでは二項演算子を見てきましたが、NOT演算を表す単項演算子もあるんですね〜。それが`~`！条件の真偽が逆転します。値がbooleanのカラムでフィルタするときによく使うんだそうですね！

たとえばこんなデータがあるとしましょう。

In [59]:
data = [{'Name':'John', 'Survived':True},
        {'Name':'Emily', 'Survived':False},
        {'Name':'Ben', 'Survived':True}]
df = pd.DataFrame(data)
df

Name  Survived
0   John      True
1  Emily     False
2    Ben      True

生存者だけでフィルタしてみます。

In [61]:
df[df['Survived']]

Name  Survived
0  John      True
2   Ben      True

さて、`Survived == False`のときはどうでしょう？そんな時に`~`を使うんですね〜。それっ。

In [62]:
df[~df['Survived']]

Name  Survived
1  Emily     False

ところでですね、勘のいい方はすでにお気づかもしれません。さっきのタイタニックのデータフィルタしたDataFrameをもう一度見てみましょうか。

In [63]:
df = pd.read_csv('data/titanic/train.csv')
df[df['Survived'] == 1]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[342 rows x 12 columns]

わかりましたか？わかりませんかね？実はですね、indexがとびとびになってるところがあるんですね〜。これは、元のDataFrameのindexが保存されてるわけですね。

これに対してindexを振り直したいというときは`df.reset_index()`を使いましょう。

In [64]:
df.reset_index()

index  PassengerId  Survived  Pclass  \
0        0            1         0       3   
1        1            2         1       1   
2        2            3         1       3   
3        3            4         1       1   
4        4            5         0       3   
..     ...          ...       ...     ...   
886    886          887         0       2   
887    887          888         1       1   
888    888          889         0       3   
889    889          890         1       1   
890    890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 13 columns]

そうすると、元のindexが`index`カラムとして保存されつつ、新たにindexが振られるわけですね〜〜〜。

もひとつついでに、別のカラムをindexにしたい場合は`df.set_index()`を使います。`Name`カラムをindexにしてみましょう。

In [65]:
df = pd.read_csv('data/titanic/train.csv')
df.set_index('Name')

PassengerId  Survived  \
Name                                                                        
Braund, Mr. Owen Harris                                       1         0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...            2         1   
Heikkinen, Miss. Laina                                        3         1   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  4         1   
Allen, Mr. William Henry                                      5         0   
...                                                         ...       ...   
Montvila, Rev. Juozas                                       887         0   
Graham, Miss. Margaret Edith                                888         1   
Johnston, Miss. Catherine Helen "Carrie"                    889         0   
Behr, Mr. Karl Howell                                       890         1   
Dooley, Mr. Patrick                                         891         0   

                                                    Pclass     Sex   Age  \
Name                                                                       
Braund, Mr. Owen Harris                                  3    male  22.0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...       1  female  38.0   
Heikkinen, Miss. Laina                                   3  female  26.0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             1  female  35.0   
Allen, Mr. William Henry                                 3    male  35.0   
...                                                    ...     ...   ...   
Montvila, Rev. Juozas                                    2    male  27.0   
Graham, Miss. Margaret Edith                             1  female  19.0   
Johnston, Miss. Catherine Helen "Carrie"                 3  female   NaN   
Behr, Mr. Karl Howell                                    1    male  26.0   
Dooley, Mr. Patrick                                      3    male  32.0   

                                                    SibSp  Parch  \
Name                                                               
Braund, Mr. Owen Harris                                 1      0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...      1      0   
Heikkinen, Miss. Laina                                  0      0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)            1      0   
Allen, Mr. William Henry                                0      0   
...                                                   ...    ...   
Montvila, Rev. Juozas                                   0      0   
Graham, Miss. Margaret Edith                            0      0   
Johnston, Miss. Catherine Helen "Carrie"                1      2   
Behr, Mr. Karl Howell                                   0      0   
Dooley, Mr. Patrick                                     0      0   

                                                              Ticket     Fare  \
Name                                                                            
Braund, Mr. Owen Harris                                    A/5 21171   7.2500   
Cumings, Mrs. John Bradley (Florence Briggs Tha...          PC 17599  71.2833   
Heikkinen, Miss. Laina                              STON/O2. 3101282   7.9250   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  113803  53.1000   
Allen, Mr. William Henry                                      373450   8.0500   
...                                                              ...      ...   
Montvila, Rev. Juozas                                         211536  13.0000   
Graham, Miss. Margaret Edith                                  112053  30.0000   
Johnston, Miss. Catherine Helen "Carrie"                  W./C. 6607  23.4500   
Behr, Mr. Karl Howell                                         111369  30.0000   
Dooley, Mr. Patrick                                           370376   7.7500   

                                                   Cabin Embarked  
Name                                                  

indexを貼り直すってことはよくあるみたいです。でも、せっかく覚えたんですが、indexをセットすることはあんまりないみたいですね〜〜〜。

というわけで、今日はDataFrameからいろんな条件でデータをフィルタして抽出してみました。今日も勉強になりましたね〜〜〜。ではでは、また〜〜〜！